### DISCUSSION TITLE
[A-L] (23 luglio 2024) Esercizio 4

### DISCUSSION THREAD

**Discussion by Elia-Belli** [sapienzastudentsnetwork/progettazione-di-algoritmi#113] [Esami A-L]
**Esercizio 4**: sia $A$ un array di interi di lunghezza $n$ . Una _montagna_ in $A$ di lunghezza $j$ è un indice $i$ tale che:
- $j \leq i \leq n-j$
- $A[i-j] \leq A[i-j+1] \leq A[i-j+2] \leq \dots \leq A[i-1] \leq A[i]$
- $A[i] \geq A [i+1] \geq A[i+2] \geq \dots \geq A[i+j]$

Dare lo pseudocodice di un algoritmo $O(n)$ per trovare la lunghezza massima di una montagna in $A$ .

**Esempio:**
<p align = center>
<img width = 300 src =https://github.com/user-attachments/assets/27c54e1c-7fe8-4d79-abaf-f1378fb4661c>
</p>

**Comment by i0nut02** [comment]
```py
def es(A):
    n = len(A)
    desc_left = [0] * n
    desc_right = [0] * n

    for i in range(1, n):
        if A[i] >= A[i-1]:
            desc_right[i] = 1 + desc_right[i-1]
    
    for i in range(n - 2, -1, -1):
        if A[i] >= A[i+1]:
            desc_left[i] = 1 + desc_left[i+1]
    
    res = 1

    for i in range(n):
        res = max(res, 1 + 2 * min(desc_left[i], desc_right[i]))
    
    return res
```

**Comment by CiottoloMaggico** [comment]
@i0nut02 Credo che con lunghezza non intenda la lunghezza effettiva della montagna ma solo il valore j che crea la montagna più lunga, comunque l'algoritmo non cambia 
![immagine](https://github.com/user-attachments/assets/48126b1c-207c-4f5e-94f7-1f066a7a4496)

**Reply by i0nut02** [reply]
si ha ragione, secondo la definizione data, nel tuo algo aggiorna ris = L[i] to ris = min(L[i], R[i]).
Inoltre della definizione di L[i] e R[i] devi stare attento quando i = 0 e i = n -1 o in generale al primo e ultimo indice

**Comment by FeddyLix17** [comment]
È possibile dare lo pseudocodice di un algoritmo $\Large O(n)$ utilizzando l'approccio [a due puntatori](https://www.reddit.com/r/leetcode/comments/18g9383/twopointer_technique_an_indepth_guide_concepts/).

Per ogni inizio di montagna identificato, si ricaveranno i rispettivi indici di inizio salita e fine discesa (la cui differenza rappresenta la lunghezza della montagna corrente).

Verrà tenuta traccia della lunghezza massima corrente durante tutto lo scorrimento dell'array.

```python3
def es4(A):
    n = len(A)
    if n < 3:
        return 0
    
    max_len = 0
    start = 0
    i = 0
    
    while i < n - 1:
        # salita
        while i < n - 1 and A[i] <= A[i + 1]:
            i += 1
        
        picco = i
        
        # discesa
        while i < n - 1 and A[i] >= A[i + 1]:
            i += 1
        
        # verifica della valitidà della sequenza corrente come una montagna
        if (picco > start and i > picco) or (A[start] == A[picco] == A[i] and i > start):
            # in caso di riscontro positivo
            max_len = max(max_len, i - start + 1)
        
        # lettura nuova potenziale montagna
        start = i
        i += 1
    
    return max_len
```

Il singolo scorrimento dell'array mantiene il costo dell'algoritmo $\Large O(n)$.

Questo esercizio risulta essere un problema famoso denominato [Longest Mountain in Array](https://leetcode.com/problems/longest-mountain-in-array/) (con la sola differenza che in quest'ultimo gli elementi della montagna devono essere strettamente minori o maggiori a seconda della situazione).

**Reply by i0nut02** [reply]
in questo caso il costo e' $O(n^{2})$ considerando i 2 cicli while, considera un array come [2] * n, in questo caso per ogni i andrai a vedere tutti gli elementi a destra e a sinistra di i, il che implica che ogni elemento e' visitato n-1 volte all'interno dei cicli while il che implica il costo detto precedentemente.

**Reply by FeddyLix17** [reply]
i cicli while fanno parte del blocco if, nel quale si entrerà solo di fronte al picco di una montagna (di lunghezza >= 3).

Per struttura stessa del problema, non possono esistere montagne con picchi diversi che abbiano elementi in comune.

Nel caso in cui ci sia un unica montagna lunga len(A), con picco a indice len(A)//2, saranno state effettuate 2n iterazioni dal ciclo for (combinando le visite fatte dal ciclo for e dai 2 cicli while, ogni elemento dell'array sarà stato visitato esattamente 2 volte).

In tutti gli altri casi, il numero di iterazioni effettuate dal ciclo for sarà $\Large [n-1,\ 2n - 1]$.

**Reply by i0nut02** [reply]
in realta' con l'array [2, 2, 2, 2, 2] tu entri sempre nel if precisamente 3 e definendo $n = len(A)$ tu entrerai nell'if esattamente $n - 2$ volte nel caso in tutti gli elementi dell'array sono uguali.

Ad esempio tu hai detto che saranno visitati esattamente 2 volte ma qui sono visitati gia' piu' di 2 volte il che ti dovrebbe far capire la complessita'.

Tipo per
- i = 1 entri nell'if e visiti tutto A
- i = 2 entri nell'if e visiti tutto A
-  i = 3 entri nell'if e visiti tutto A

e ovviamente piu' e' grande l'array e con elementi tutti uguali, fai sempre almeno n - 2 visite per ogni elemento.

**Reply by FeddyLix17** [reply]
@i0nut02 ti ringrazio per l'ulteriore chiarimento, ho modificato l'algoritmo per poter sia gestire il caso di elementi tutti uguali nell'array che per garantirne il costo lineare.

**Reply by i0nut02** [reply]
ora effettivamente ha costo $O(n)$ ma il problema e' che non trovi il lato piu' corto, inoltre il picco non puo' rappresentare la parte centrare siccome nel caso [2, 2, 2, 2, 2] sara' l'ultimo elemento che ha 0 elementi a destra, e la risposta sarebbe erratta
